Plutarch/Natural.hs

```diff
@@ -0,0 +1,216 @@
+-- This prevents errors on 'instance IntegralDomain PInteger PNatural'
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Plutarch.Natural (
+  PNatural (..),
+  pscaleNat,
+  ppowNat,
+  pnatToInt,
+  pnatFromInt,
+) where
+
+import Plutarch (PlutusType (..), punsafeCoerce)
+import Plutarch.Bool (
+  PEq ((#==)),
+  POrd ((#<), (#<=)),
+  pif,
+ )
+import Plutarch.Builtin (
+  PAsData,
+  PIsData (..),
+  pasInt,
+  pforgetData,
+ )
+import Plutarch.Integer (PInteger)
+import Plutarch.Lift (pconstant)
+import Plutarch.Maybe (PMaybe (PJust, PNothing))
+import Plutarch.Numeric (
+  PAdditiveGroup ((#-)),
+  PAdditiveHemigroup ((#^-)),
+  PAdditiveMonoid (pzero),
+  PAdditiveSemigroup ((#+)),
+  PEuclideanClosed (pdivMod),
+  PIntegralDomain (pabs, paddExtend, pprojectAbs, prestrictMay),
+  PMultiplicativeMonoid (pone),
+  PMultiplicativeSemigroup ((#*)),
+  pdiv,
+  peven,
+  pnegate,
+ )
+import Plutarch.Pair (PPair (PPair))
+import Plutarch.Prelude
+
+newtype PNatural (s :: S) = PNatural (Term s PInteger)
+
+instance PIsData PNatural where
+  pfromData x =
+    phoistAcyclic
```

Comment:
You can just use `pfromData` for integers here without all this complex logic. The argument passed to this function is `PAsData PNatural`, so we already know it's correct. No need to check anything.